File: products/LoanHub/apps/backend/database/schemas/notification.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from database.models.enums import NotificationType


class NotificationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    actor_user_id: UUID | None = None
    company_id: UUID | None = None
    branch_id: UUID | None = None

    title: str
    message: str
    notification_type: NotificationType
    event_type: str
    action: str
    entity_type: str | None = None
    entity_id: str | None = None
    action_url: str | None = None
    icon: str | None = None
    priority: str
    data: dict[str, Any] = Field(default_factory=dict)

    is_read: bool
    read_at: datetime | None = None
    is_archived: bool
    archived_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class NotificationListRead(BaseModel):
    items: list[NotificationRead]
    total: int
    unread_count: int
    page: int
    page_size: int


class NotificationUnreadCountRead(BaseModel):
    unread_count: int

File: products/LoanHub/apps/frontend/types/notification.ts
export type NotificationType =
    | "loan_request"
    | "access_request"
    | "offer"
    | "payment"
    | "subscription"
    | "system";

export type AppNotification = {
    id: string;
    user_id: string;
    actor_user_id: string | null;
    company_id: string | null;
    branch_id: string | null;
    title: string;
    message: string;
    notification_type: NotificationType;
    event_type: string;
    action: string;
    entity_type: string | null;
    entity_id: string | null;
    action_url: string | null;
    icon: string | null;
    priority: string;
    data: Record<string, unknown>;
    is_read: boolean;
    read_at: string | null;
    is_archived: boolean;
    archived_at: string | null;
    created_at: string;
    updated_at: string;
};

export type NotificationListResponse = {
    items: AppNotification[];
    total: number;
    unread_count: number;
    page: number;
    page_size: number;
};
